### ISSUE TITLE
log_softmax get inf with M1 MPS backend

### ISSUE THREAD

**Issue by sailist** [pytorch/pytorch#94043]
### 🐛 Describe the bug

When using `log_softmax`, I got different results on different backend, two samples are provided:
```
from torch import tensor
a = tensor([[ -16.9397,   74.5907, -245.5731,   57.9282,  -59.7757,   26.2957,
          -27.3786,   71.2639,   57.6391,   54.0626],
        [ -13.8363,   62.5153, -206.0240,   48.6378,  -51.2763,   21.7313,
          -22.6663,   59.4996,   47.2677,   47.7668]], device='mps:0')
print(a.log_softmax(dim=-1))
print(a.cpu().log_softmax(dim=-1))
```
I got the following results:
```
tensor([[       -inf,     -0.0353,        -inf,    -16.6978,        -inf,
            -48.3303,        -inf,     -3.3621,    -16.9869,    -20.5634],
        [   -76.3995,     -0.0478,        -inf,    -13.9254,        -inf,
            -40.8318,    -85.2295,     -3.0636,    -15.2955,    -14.7963]],
       device='mps:0')
tensor([[   -91.5657,     -0.0353,   -320.1991,    -16.6978,   -134.4017,
            -48.3303,   -102.0046,     -3.3621,    -16.9869,    -20.5634],
        [   -76.3995,     -0.0478,   -268.5872,    -13.9254,   -113.8394,
            -40.8319,    -85.2295,     -3.0636,    -15.2955,    -14.7964]])
```

### Versions

```
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6.2 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.21.3
Libc version: N/A

Python version: 3.9.12 (main, Mar 26 2022, 15:45:34)  [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-12.6.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] pytorch-pretrained-vit==0.0.7
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchtext==0.10.0
[pip3] torchvision==0.14.1
[pip3] vit-pytorch==0.25.6
[conda] Could not collect
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by malfet** [comment]
Hmm, looks like a simple division accuracy problem, i.e.:
```
% python -c "import torch;print(torch.tensor([4.3e-08, 2.4e32], device='mps')/ 2.5e32)" 
tensor([0.0000, 0.9600], device='mps:0')
% python -c "import torch;print(torch.tensor([4.3e-08, 2.4e32], device='cpu')/ 2.5e32)"
tensor([1.7200e-40, 9.6000e-01])
```

**Comment by sailist** [comment]
> Hmm, looks like MPS like many other GPUs does not support denorms, i.e. see following example:
> 
> ```
> % python -c "import torch;print(torch.tensor([4.3e-08, 2.4e32], device='mps') / 2.5e32)" 
> tensor([0.0000, 0.9600], device='mps:0')
> % python -c "import torch;print(torch.tensor([4.3e-08, 2.4e32], device='cpu') / 2.5e32)"
> tensor([1.7200e-40, 9.6000e-01])
> ```
> 
> 1.72e-40 can not be represented as normalized 32-bit float, but can be represented via denorm. To cite [wikipedia](https://en.wikipedia.org/wiki/Single-precision_floating-point_format#Exponent_encoding):
> 
> > The minimum positive normal value is 2−126≈1.18∗10−38

You are right, does that mean there is nothing I can do to fix this right now?

**Comment by malfet** [comment]
Grabbing for myself, as this sounds like fun.
Both CPU and GPU implementations of `log_softmax` use following formula: $logsoftmax(a_j) = a_j-log(\sum{e^{a_i}})$ , where MPS implementation takes a literal $logsoftmax(a_j) = log\frac{e^{a_j}}{\sum{e^{a_i}}}$, which probably have lower accuracy.

Compare:
https://github.com/pytorch/pytorch/blob/d2b82feb41552e93efdd8f876b7a938d9657a1d6/aten/src/ATen/native/mps/operations/Activation.mm#L305-L309
With:
https://github.com/pytorch/pytorch/blob/d2b82feb41552e93efdd8f876b7a938d9657a1d6/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L967-L977

**Comment by sailist** [comment]
Good finding, for others who meet these problem, you can workaround it by writing your own `log_softmax` function:
```
def log_softmax(x,dim=-1):
    c = x.max(dim=dim,keepdim=True).values
    logsumexp = torch.log(torch.exp(x - c).sum(dim=dim,keepdim=True))
    return x - c - logsumexp

def cross_entropy(x,dim=-1):
    return F.nll_loss(logits.log_softmax(dim=dim),labels)

a = tensor([[ -16.9397,   74.5907, -245.5731,   57.9282,  -59.7757,   26.2957,
          -27.3786,   71.2639,   57.6391,   54.0626],
        [ -13.8363,   62.5153, -206.0240,   48.6378,  -51.2763,   21.7313,
          -22.6663,   59.4996,   47.2677,   47.7668]], device='mps')

print(torch.abs(log_softmax(a).cpu() - a.cpu().log_softmax(dim=-1)) < 1e-6)
>>> tensor([[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True],
        [ True,  True,  True,  True,  True, False,  True,  True,  True,  True]])
```

**Comment by AlexRoar** [comment]
While this is nice that it can be solved by writing your own `log_softmax`, I'd prefer this to be fixed in the `pytorch` itself. I recently faced the same issue and it's devastating cause it ruins the whiole training. Also, it's hareder to use your own function in pretrained nns.

@sailist, could you please reopen this?

**Comment by sailist** [comment]
Sure, I forgot that issue can be used to remind developers to upgrade code.
